include/circt/Dialect/Scheduling/Algorithms/ASAPScheduler.h

```diff
@@ -0,0 +1,46 @@
+//===- ASAPScheduler.h - As-soon-as-possible list scheduler -----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines an acyclic, as-soon-as-possible list scheduler.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_SCHEDULING_ALGORITHMS_ASAPSCHEDULER_H
+#define CIRCT_DIALECT_SCHEDULING_ALGORITHMS_ASAPSCHEDULER_H
+
+#include "circt/Dialect/Scheduling/Scheduler.h"
+
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/DenseMap.h"
+
+namespace circt {
+namespace sched {
+
+class ASAPScheduler : public SchedulerBase {
+private:
+  llvm::SetVector<Operation *> operations;
+  llvm::DenseMap<Operation *, SmallVector<Operation *, 4>> dependences;
+  llvm::DenseMap<Operation *, OperatorInfoAttr> operators;
+  llvm::DenseMap<Operation *, unsigned> startTimes;
```

Comment:
I was wondering this as well.

Reply:
Yes, different algorithms may desire different data structures. For example, when the ASAP scheduler in this PR tries to schedule an operation, it needs to query that op's predecessors, so it made sense to store the dependence info in a `map<Op*, list<Op*>`. However, if a scheduler internally constructs a linear program (where each dependence is mapped to a linear constraint), it would be more convenient to iterate over a list of dependence edges.

Also, the "interface" (I mean the `register...` methods) currently is as general as possible (and there should be simpler convenience variants of the methods in the future), but not every algorithm will support the generality. For example, a non-modulo scheduler does not need to store the distance information, and an algorithm that doesn't choose from multiple operator types only needs to store a single one.

Therefore I'd argue that the choice of data structures should be up to the scheduler (or to a superclass that represents a group of similar algorithms).